src/borg/testsuite/hashindex.py

```diff
@@ -486,87 +477,57 @@ def write_empty(self, key):
     def write_deleted(self, key):
         self.write_entry(key, 0xFFFFFFFE, 0, 0)
 
+    def compare_indexes(self, idx1, idx2):
+        for k, v in idx1.iteritems():
+            assert v == idx2[k]
```

Comment:
If I write `assert idx1[k] == idx2[k]` here I get a key not found error.  Any use of `idx1[k]` gives this error.  Is there some reason I can't look up a key while iterating over the hashtable?  I'm just mentioning it in case it indicates a bug somewhere.

Reply:
I don't remember any reason right now, will have a deeper look later.

Reply:
idx1 is an artificially (and not necessarily correctly) constructed HT, idx2 ("cpt") is constructed the same way, but then compacted.

When doing random lookups, the lookup algorithm relies on the HT being set up **correctly** (and gives up searching for correct key once it finds an empty bucket).

This makes lookups in the non-compact table fail, while lookups in the compact table still succeed because there are no empty buckets.

Iteration just goes over everything (skipping empty/deleted), so does not have issues with bad HTs.

Guess this could be copied to a comment here.

Reply:
Oh, I figured it out.  It has nothing to do with the iterator, but is simply because the entries aren't located where they should be.  I've added comments explaining this.

Reply:
Guess we found it in parallel. :-)